### ISSUE TITLE
Signals cause reprojection errors on empty datasets

### ISSUE THREAD

**Issue by nyurik** [vega/vega#1099]
This is the minimal example I could come up with.  It works fine in the editor (shows as a blank area, no console errors), but as soon as you click it, console shows this error.

Note that editor does not show any of them to the developer.

```
ERROR TypeError: e.mark is undefined
Stack trace:
nr.dirty@https://vega.github.io/editor/static/js/main.26731473.js:9:25002
fe.dirty@https://vega.github.io/editor/static/js/main.26731473.js:51:18818
I.transform/<@https://vega.github.io/editor/static/js/main.26731473.js:51:6012
le@https://vega.github.io/editor/static/js/main.26731473.js:2:4611
_e.visit@https://vega.github.io/editor/static/js/main.26731473.js:4:10323
I.transform@https://vega.github.io/editor/static/js/main.26731473.js:51:5958
ke.evaluate@https://vega.github.io/editor/static/js/main.26731473.js:4:12567
ke.run@https://vega.github.io/editor/static/js/main.26731473.js:4:12428
C@https://vega.github.io/editor/static/js/main.26731473.js:4:2588
fe.run@https://vega.github.io/editor/static/js/main.26731473.js:51:18494
s@https://vega.github.io/editor/static/js/main.26731473.js:51:10306
Jn.fire@https://vega.github.io/editor/static/js/main.26731473.js:9:23520
Jn.click@https://vega.github.io/editor/static/js/main.26731473.js:9:22918
Jn.initialize/</<@https://vega.github.io/editor/static/js/main.26731473.js:9:22186
 vega-util.js:132
```

```json
{
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "padding": 0,
  "width": 400,
  "height": 500,
  "autosize": "none",
  "data": [
    {
      "name": "countries",
      "url": "https://vega.github.io/editor/data/world-110m.json",
      "format": {"type": "topojson", "feature": "countries"},
      "transform": [
        {
          "type": "formula",
          "expr": "geoCentroid('projection', datum)",
          "as": "centroid"
        },
        {"type": "filter", "expr": "false"}    ///  <<----  works without this filter
      ]
    }
  ],
  "marks": [
    {
      "type": "shape",
      "from": {"data": "countries"},
      "transform": [{"type": "geoshape", "projection": "projection"}]
    }
  ],
  "signals": [
    {
      "name": "zoom",
      "value": 0,
      "on": [{"events": {"type": "click"}, "update": "(zoom+1)%3"}]
    },
    {"name": "latitude", "value": 0},
    {"name": "longitude", "value": 0}
  ],
  "projections": [
    {
      "name": "projection",
      "type": "mercator",
      "scale": {"signal": "256*pow(2,zoom)/2/PI"}
    }
  ]
}
```

**Comment by jheer** [comment]
Thanks! The error here was due to a subtle bug in the dataflow engine that caused tuples to be propagated in an improper way. This has been fixed and released in [vega-dataflow v3.0.4](https://github.com/vega/vega-dataflow/releases/tag/v3.0.4), and will be included in the next release of Vega.

**Comment by nyurik** [comment]
@jheer thanks for fixing it so fast!
Should there be a unit test for this? I'm worried as time goes by there could be more and more such bugs that are hard to test for, as other parts of Vega libs change.
